tools/pyboard.py: Add fs_listdir/fs_readfile/fs_writefile and a parse kwarg to eval.

jimmo opened the pull request:
This is to make it easier to use pyboard.py as a library. The existing eval/fs_ls/fs_cat are really only useful right now when using it from the command line.

- fs_listdir returns a list of tuples, in the same format as os.ilistdir().
- fs_readfile returns the contents of a file as a bytes object.
- fs_writefile allows writing a bytes object to a file.
- eval(..., parse=True) will evaluate the result into a Python object. This works for any Python literal (i.e. that can be understood by `ast.parse_literal`).

```
>>> pyb.eval("1+1", parse=True)
2
>>> pyb.eval("{'a': '\x00'}", parse=True)
{'a': '\x00'}
```

fs_listdir in particular might be useful to use in mpremote for future recursive operations. Another example is that mpremote/mip does `int(pyb.eval("getattr(sys.implementation, '_mpy', 0) & 0xFF").decode())` which could now become `pyb.eval("getattr(sys.implementation, '_mpy', 0) & 0xFF", parse=True)`.

_This work was funded through GitHub Sponsors._

jimmo committed:
tools/pyboard.py: Add fs_{listdir,readfile,writefile,stat}.

These are for working with the filesystem when using pyboard.py as a
library, rather than at the command line.

- fs_listdir returns a list of tuples, in the same format as os.ilistdir().
- fs_readfile returns the contents of a file as a bytes object.
- fs_writefile allows writing a bytes object to a file.
- fs_stat returns an os.statresult.

All raise FileNotFoundError (or OSError(ENOENT) on Python 2) if the file is
not found (or PyboardError on other errors).

Updated fs_cp and fs_get to use fs_stat to compute file size.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tools/pyboard.py: Add parse kwarg to eval.

This is useful when using pyboard.py as a library rather than at the
command line.

    pyb.eval("1+1") --> b"2"
    pyb.eval("{'a': '\x00'}") --> b"{'a': '\\x00'}"

Now you can also do

    pyb.eval("1+1", parse=True) --> 2
    pyb.eval("{'a': '\x00'}", parse=True) --> {'a': '\x00'}

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

peterhinch commented:
This looks good. An enhancement for `mpremote` that has been mentioned is an `rsync` command similar to that in `rshell`. This uses modification time to determine whether a file needs to be updated on the target. A means of accessing `os.stat(file_path)` would facilitate this.

jimmo commented:
> A means of accessing `os.stat(file_path)` would facilitate this.

Good idea! I've added `fs_stat` that returns an `os.stat_result` exactly like os.stat would. This is actually useful in pyboard.py itself too, and definitely for mpremote.

Also updated `fs_listdir` to return a named tuple, and all methods to raise FileNotFoundError rather than PyboardError when the board returns file not found.

jimmo commented:
I assume we're still maintaining python2 compatibility for pyboard.py, so I will update the PR to fix that.

jimmo committed:
tools/pyboard.py: Fix Python 2 compatibility.

In Python 2, serial.read()[0] a string, not int. Use struct.unpack to do
this instead.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> I assume we're still maintaining python2 compatibility for pyboard.py, so I will update the PR to fix that.

Turns out the current code in master wasn't working in python2 either, but added a commit to fix that too.

dpgeorge commented:
Thank you!
